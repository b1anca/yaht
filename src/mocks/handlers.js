import { rest } from "msw";

export const handlers = [
  rest.post(
    `${process.env.REACT_APP_YAHT_API_URL}/auth/login`,
    (req, res, ctx) => {
      const { email } = req.body;

      return res(
        ctx.json({ token: `token-${email}`, name: "user name", email }),
        ctx.status(200)
      );
    }
  ),
  rest.post(`${process.env.REACT_APP_YAHT_API_URL}/users`, (req, res, ctx) => {
    const { email } = req.body;

    return res(ctx.json({ id: 1, name: "user name", email }), ctx.status(200));
  }),
  rest.get(`${process.env.REACT_APP_YAHT_API_URL}/users/me`, (_, res, ctx) => {
    return res(ctx.status(401));
  }),
  rest.get(`${process.env.REACT_APP_YAHT_API_URL}/habits`, (_, res, ctx) => {
    return res(
      ctx.json([
        {
          id: 1,
          name: "Reading",
          color: "#75ab74",
          tasks: [{ id: 11, completed_at: "2023-06-02T03:00:00.000Z" }],
          description: "Reading 10 pages",
          overall_progress: "90.0",
        },
      ]),
      ctx.status(200)
    );
  }),
];
