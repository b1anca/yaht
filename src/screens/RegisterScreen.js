import React, { useEffect, useState } from "react";
import { useForm } from "react-hook-form";
import { useDispatch, useSelector } from "react-redux";
import { useNavigate } from "react-router-dom";
import Alert from "../components/Alert";
import { registerUser } from "../features/auth/authActions";
import Input from "../components/Input";
import Button from "../components/Button";
import NavLink from "../components/NavLink";
import Header from "../components/Header";

const RegisterScreen = () => {
  const [customError, setCustomError] = useState(null);
  const { loading, userInfo, error, success } = useSelector(
    (state) => state.auth
  );
  const dispatch = useDispatch();
  const { register, handleSubmit } = useForm();
  const navigate = useNavigate();

  useEffect(() => {
    if (userInfo) navigate("/habits");
    if (success) navigate("/login");
  }, [navigate, userInfo, success]);

  const submitForm = (data) => {
    if (data.password !== data.password_confirmation) {
      setCustomError("Password mismatch");
      return;
    }
    data.email = data.email.toLowerCase();

    dispatch(registerUser(data));
  };

  return (
    <>
      <div className="relative h-screen flex flex-1 flex-col overflow-hidden">
        <Header />
        <div className="relative flex flex-1 flex-col items-center justify-center">
          <form
            className="w-full max-w-sm bg-slate-300/5 rounded px-5 py-12"
            onSubmit={handleSubmit(submitForm)}
          >
            {error && <Alert>{error}</Alert>}
            {customError && <Alert>{customError}</Alert>}
            <Input type="text" required {...register("name")} label="Name" />
            <Input type="email" required {...register("email")} label="Email" />
            <Input
              type="password"
              required
              {...register("password")}
              label="Password"
            />
            <Input
              type="password"
              required
              {...register("password_confirmation")}
              label="Confirm password"
            />
            <Button primary type="sumit" loading={loading} className="w-full">
              Register
            </Button>
          </form>
        </div>
        <footer className="relative shrink-0 mb-8">
          <div className="space-y-4 text-sm sm:flex sm:items-center sm:justify-center sm:space-x-4 sm:space-y-0">
            <p className="text-center sm:text-left">Already have an account?</p>
            <NavLink secondary to="/login">
              Sign in
            </NavLink>
          </div>
        </footer>
      </div>
    </>
  );
};

export default RegisterScreen;
