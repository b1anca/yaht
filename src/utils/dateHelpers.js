export const getDaysInMonth = (month, year) => {
  const date = new Date(year, month, 1);
  const days = [];
  while (date.getMonth() === month) {
    days.push(new Date(date));
    date.setDate(date.getDate() + 1);
  }
  return days;
};

export const formatDate = (date) => {
  const monthNames = [
    "Jan",
    "Feb",
    "Mar",
    "Apr",
    "May",
    "Jun",
    "Jul",
    "Aug",
    "Sep",
    "Oct",
    "Nov",
    "Dec",
  ];
  const day = date.getDate();
  const monthIndex = date.getMonth();
  const year = date.getFullYear();

  return day + " " + monthNames[monthIndex] + " " + year;
};

export const areDatesEqual = (date1, date2) => {
  return (
    date1.getFullYear() === date2.getFullYear() &&
    date1.getMonth() === date2.getMonth() &&
    date1.getDate() === date2.getDate()
  );
};
