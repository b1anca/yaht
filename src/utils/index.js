export const getRangeFromViewportWidth = () => {
  const viewportWidth = window.innerWidth;

  if (viewportWidth <= 390) {
    return 3;
  } else if (viewportWidth <= 500) {
    return 4;
    // sm
  } else if (viewportWidth <= 640) {
    return 6;
    // md
  } else if (viewportWidth <= 768) {
    return 8;
    // lg
  } else if (viewportWidth <= 1024) {
    return 9;
  } else {
    return 10;
  }
};
