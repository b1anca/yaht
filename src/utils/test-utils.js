import React from "react";
import { render } from "@testing-library/react";
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";
import userEvent from "@testing-library/user-event";
import { setupStore } from "../app/store";

export function renderWithProviders(
  ui,
  {
    preloadedState = {},
    store = setupStore(preloadedState),
    ...renderOptions
  } = {}
) {
  // eslint-disable-next-line react/prop-types
  function Wrapper({ children }) {
    return (
      <Provider store={store}>
        <BrowserRouter>{children}</BrowserRouter>
      </Provider>
    );
  }
  return {
    store,
    user: userEvent.setup(),
    ...render(ui, { wrapper: Wrapper, ...renderOptions }),
  };
}
