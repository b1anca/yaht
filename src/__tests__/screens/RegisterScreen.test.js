import React from "react";
import { rest } from "msw";
import { fireEvent, screen } from "@testing-library/react";
import { renderWithProviders } from "../../utils/test-utils";
import { server } from "../../mocks/server";
import RegisterScreen from "../../screens/RegisterScreen";

const mockedNavigate = jest.fn();

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: () => mockedNavigate,
}));

describe("RegisterScreen", () => {
  it("should render correctly", () => {
    const { asFragment } = renderWithProviders(<RegisterScreen />);

    expect(asFragment()).toMatchSnapshot();
  });

  it("should navigate to login after form submit", async () => {
    const { user } = renderWithProviders(<RegisterScreen />);
    const nameInput = screen.getByLabelText("name");
    const emailInput = screen.getByLabelText("email");
    const passwordInput = screen.getByLabelText("password");
    const passwordConfirmInput = screen.getByLabelText("password_confirmation");

    fireEvent.change(nameInput, { target: { value: "username" } });
    fireEvent.change(emailInput, { target: { value: "user@example.com" } });
    fireEvent.change(passwordInput, { target: { value: "examplePassword" } });
    fireEvent.change(passwordConfirmInput, {
      target: { value: "examplePassword" },
    });
    await user.click(
      screen.getByRole("button", { name: "Create account", type: "submit" })
    );

    expect(mockedNavigate).toHaveBeenCalledWith("/login");
  });

  it("should show error message on server error", async () => {
    server.use(
      rest.post(
        `${process.env.REACT_APP_YAHT_API_URL}/users`,
        (_, res, ctx) => {
          return res(ctx.json({ error: "unauthorized" }), ctx.status(401));
        }
      )
    );
    const { user } = renderWithProviders(<RegisterScreen />);

    user.click(screen.getByRole("button", { name: "Create account" }));
    await screen.findByRole("alert");

    expect(screen.getByRole("alert")).toHaveTextContent(
      "Request failed with status code 401"
    );
  });
});
