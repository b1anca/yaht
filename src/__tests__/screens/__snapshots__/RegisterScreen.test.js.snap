// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegisterScreen should render correctly 1`] = `
<DocumentFragment>
  <div
    class="relative h-screen flex flex-1 flex-col overflow-hidden"
  >
    <header
      class="sticky top-0 z-40 w-full backdrop-blur flex-none transition-colors duration-500 border-b border-slate-900/10 text-sm font-semibold leading-6 bg-slate-300/5"
    >
      <div
        class="max-w-7xl mx-auto"
      >
        <div
          class="py-4 mx-4"
        >
          <div
            class="relative flex items-center"
          >
            <div
              class="relative hidden lg:flex items-center ml-auto"
            >
              <a
                class="text-sm whitespace-nowrap py-2 px-6 rounded-full inline-flex justify-center rounded-lg font-semibold py-2.5 px-4 hover:text-slate-200"
                href="/login"
              >
                Sign in
              </a>
              <a
                class="text-sm whitespace-nowrap py-2 px-6 rounded-full inline-flex justify-center rounded-lg font-semibold py-2.5 px-4 bg-slate-300 text-slate-900 hover:bg-slate-200 w-full"
                href="/register"
              >
                Register
              </a>
            </div>
          </div>
        </div>
      </div>
    </header>
    <div
      class="relative flex flex-1 flex-col items-center justify-center"
    >
      <form
        class="w-full max-w-sm bg-slate-300/5 rounded px-5 py-12"
      >
        <div
          class="mb-6"
        >
          <label
            class="block text-sm font-semibold leading-6"
            for="text"
          >
            Name
          </label>
          <input
            aria-label="name"
            class="bg-transparent mt-2 appearance-none text-slate-100 rounded-md block w-full px-3 h-10 shadow-sm sm:text-sm focus:outline-none placeholder:text-slate-300 focus:ring-2 focus:ring-sky-500 ring-1 ring-slate-200"
            id="text"
            name="name"
            required=""
            type="text"
          />
        </div>
        <div
          class="mb-6"
        >
          <label
            class="block text-sm font-semibold leading-6"
            for="email"
          >
            Email
          </label>
          <input
            aria-label="email"
            class="bg-transparent mt-2 appearance-none text-slate-100 rounded-md block w-full px-3 h-10 shadow-sm sm:text-sm focus:outline-none placeholder:text-slate-300 focus:ring-2 focus:ring-sky-500 ring-1 ring-slate-200"
            id="email"
            name="email"
            required=""
            type="email"
          />
        </div>
        <div
          class="mb-6"
        >
          <label
            class="block text-sm font-semibold leading-6"
            for="password"
          >
            Password
          </label>
          <input
            aria-label="password"
            class="bg-transparent mt-2 appearance-none text-slate-100 rounded-md block w-full px-3 h-10 shadow-sm sm:text-sm focus:outline-none placeholder:text-slate-300 focus:ring-2 focus:ring-sky-500 ring-1 ring-slate-200"
            id="password"
            name="password"
            required=""
            type="password"
          />
        </div>
        <div
          class="mb-6"
        >
          <label
            class="block text-sm font-semibold leading-6"
            for="password"
          >
            Confirm password
          </label>
          <input
            aria-label="password_confirmation"
            class="bg-transparent mt-2 appearance-none text-slate-100 rounded-md block w-full px-3 h-10 shadow-sm sm:text-sm focus:outline-none placeholder:text-slate-300 focus:ring-2 focus:ring-sky-500 ring-1 ring-slate-200"
            id="password"
            name="password_confirmation"
            required=""
            type="password"
          />
        </div>
        <button
          class="text-sm button py-2 px-6 rounded-full inline-flex justify-center rounded-lg text-sm font-semibold py-2.5 px-4 bg-slate-300 text-slate-900 hover:bg-slate-200 w-full w-full"
          type="sumit"
        >
          Register
        </button>
      </form>
    </div>
    <footer
      class="relative shrink-0 mb-8"
    >
      <div
        class="space-y-4 text-sm sm:flex sm:items-center sm:justify-center sm:space-x-4 sm:space-y-0"
      >
        <p
          class="text-center sm:text-left"
        >
          Already have an account?
        </p>
        <a
          class="text-sm whitespace-nowrap py-2 px-6 rounded-full inline-flex justify-center rounded-lg font-semibold py-2.5 px-4 ring-1 ring-slate-300/10 hover:ring-slate-200/20"
          href="/login"
        >
          Sign in
        </a>
      </div>
    </footer>
  </div>
</DocumentFragment>
`;
